docs/design.md

```diff
@@ -0,0 +1,299 @@
+   follows and abort.
+
+   ```
+   next_head := new_head
+   ```
+
+4. `next_tail` is null, which means we have finished ingesting all
+   advertisements from `head` to the end of the chain. Update the state as
+   follows and start the chain walker.
+
+   ```
+   next_head = new_head
+   head = new_head
+   tail = new_head
+   ```
+
+The chain-walking algorithm loops over the following steps:
+
+1. If ` tail == last_head || tail == null`, then we finished the walk. Update
+   the state as follows:
+
+   ```
+   last_head = head
+   head = null
+   tail = null
+   ```
+
+   If `next_head != last_head` then start a new walk by updating the state as
+   follows:
+
+   ```
+   head = next_head
+   tail = next_head
+   ```
+
+2. Otherwise take a step to the next item in the chain:
+
+   1. Fetch the advertisement identified by `tail` from the index provider.
+   2. Process the metadata and entries to extract one `(PieceCID, PayloadCID)`
+      entry.
+   3. Update the `tail` field using the `PreviousID` field from the
+      advertisement.
+
+   ```
+   tail = PreviousID
+   ```
+
+#### Handling the Scale
+
+At the time of writing this document, cid.contact was tracking 322 index
+providers. From Sparks' measurements, we know there are an additional 843
+storage providers that don't advertise to IPNI. The number of storage providers
+grows over time, our system must be prepared to ingest advertisements from
+thousands of providers.
+
+Each storage/index provider produces tens of thousands to hundreds of thousands
+of advertisements (in total). The initial ingestion run will take a while to
+complete. We also must be careful to not overload the SP by sending too many
+requests.
+
+The design outlined in the previous section divides the ingestion process into
+small steps that can be scheduled and executed independently. This allows us to
+avoid the complexity of managing long-running per-provider tasks and instead
+repeatedly execute one step of the process.
+
+Loop 1: Every minute, fetch the latest provider information from cid.contact and
+update the persisted state as outlined above.
+
+Loop 2: Discover walks in progress and make one step in each walk.
+
+1. Find all provider state records where `tail != null`.
+
+2. For each provider, execute one step as described above. We can execute these
+   steps in parallel. Since each parallel job will query a different provider,
+   we are not going to overload any single provider.
+
+3. Optionally, we can introduce a small delay before the next iteration. I think
+   we won't need it because the time to execute SQL queries should create enough
+   delay.
+
+### REST API
+
+Implement the following endpoint that will be called by Spark checker nodes. The
+endpoint will sign the response using the server's private key to allow
+spark-evaluate to verify the authenticity of results reported by checker nodes:
+
+```
+GET /sample/{providerId}/{pieceCid}?seed={seed}
```

Comment:
can we add an endpoint that allows a SP to see which records there are in the DB associated to them? As per F8 Ptrk's comment a week or so back

Reply:
See the "Observability" section below. Do you think we need something different?